File: leetcode/algorithms/507_perfect_number/main.py
import math


class PerfectNumber:
    def checkPerfectNumber(self, num: int) -> bool:
        if num == 1:
            return False

        sum = 0
        for i in range(1, math.floor(num**0.5) + 1):
            print(i)
            if num % i == 0:
                sum += i + num / i

        return sum == num * 2


    # Solution
    # Solution 1: Only 5 perfect Number exist until 10^8
    def solution1(self, num: int) -> bool:
        return num in [6, 28, 496, 8128, 33550336]

    # Solution 2: Rooting Approach
    def solution2(self, num: int) -> bool:
        if num == 1:
            return False

        count = 1
        for i in range(2, int(num**0.5) + 1):
            if num % i == 0:
                count += i + num // i

        return num == count

File: leetcode/algorithms/507_perfect_number/main.dart
class PerfectNumber {
  bool checkPerfectNumber(int num) {
    int sum = 0;
    for (int i = 1; i < num / 2; i++) {
      if (num % i == 0) {
        sum += i;
      }
    }
    return sum * 2 == num;
  }

  // Solution
  bool solution(int num) {
    int sum = 0;
    for (int i = 1; i * i <= num; i++) {
      if (num % i == 0) {
        sum += i;
        if (i * i != num) {
          sum += num ~/ i;
        }
      }
    }
    return sum - num == num;
  }
}
